### PR TITLE
Add Conv2dNormActivation and Conv3dNormActivation Blocks

### LINKED ISSUE #5430: ConvNormActivation block for 3D convolutions
### 🚀 The feature

I suggest we add a `Conv3d` equivalent for the  `ConvNormActivation` block

### Motivation, pitch

This will be used in the upcoming ResNext3d implementation. 

### Alternatives

I see two possible alternatives:

1. We can re-purpose the existing method in a BC way  by adding an extra named parameter `conv layer` with default value of `Conv2D`. 

2. We can explicitly add a new class `Conv3DNormActivation`, which will be very similar to the existing one.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by oke-aditya**
Closes #5430 

Simple code to try this. It worked fine.

```
class SimpleModel(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.layer1 = Conv3dNormActivation(3, 15, stride=2, padding=2,
                                           activation_layer=torch.nn.LeakyReLU,
                                           inplace=True,
                                           bias=True)
    
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.layer1(x)
        return x

class EasyModel(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.layer1 = Conv2dNormActivation(3, 15, stride=2, padding=2,
                                           activation_layer=torch.nn.LeakyReLU,
                                           inplace=True,
                                           bias=True)
    
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.layer1(x)
        return x



if __name__ == "__main__":

    image_tensor = torch.rand((5, 3, 224, 224))
    model = EasyModel()
    out = model(image_tensor)
    print(out.size())  # torch.Size([5, 15, 113, 113])

    video_tensor = torch.rand((5, 3, 10, 224, 224))
    model = SimpleModel()
    out = model(video_tensor)
    print(out.size())  # torch.Size([5, 15, 6, 113, 113]

```

### CODE REVIEW

File: torchvision/ops/misc.py [hunk 1]
```diff
@@ -161,3 +161,62 @@ def _scale(self, input: Tensor) -> Tensor:
...
+        padding: Optional[int] = None,
+        groups: int = 1,
+        norm_layer: Optional[Callable[..., torch.nn.Module]] = torch.nn.BatchNorm3d,
+        activation_layer: Optional[Callable[..., torch.nn.Module]] = torch.nn.ReLU,
+        dilation: int = 1,
+        inplace: Optional[bool] = True,
+        bias: Optional[bool] = None,
+    ) -> None:
+
+        if padding is None:
```
Comment (datumbox) [inline torchvision/ops/misc.py:199]: There is lots of duplicate code with 2d. Should we have a base class to avoid it and both implementation call the same base? PyTorch dos something similar with [BatchNorm](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L338).

Comment (oke-aditya) [reply torchvision/ops/misc.py:199]: I thought the same. There is duplication. We can probably just create a base class convbatchNorm and then extend that for 2d and 3d ?

Comment (jdsgomes) [reply torchvision/ops/misc.py:199]: I agree, please see my proposal here: https://github.com/pytorch/vision/pull/5440

File: torchvision/ops/__init__.py [hunk 2]
```diff
@@ -52,7 +52,7 @@
     "StochasticDepth",
     "FrozenBatchNorm2d",
     "ConvNormActivation",
-    "Conv3dNormActivation",
+    "Conv2dNormActivation" "Conv3dNormActivation",
```
Comment (datumbox) [inline torchvision/ops/__init__.py:55]:
```suggestion
    "Conv2dNormActivation", "Conv3dNormActivation",
```

File: torchvision/ops/misc.py [hunk 3]
```diff
@@ -114,56 +100,125 @@ def __init__(
...
+        stride (int, optional): Stride of the convolution. Default: 1
+        padding (int, tuple or str, optional): Padding added to all four sides of the input. Default: None, in wich case it will calculated as ``padding = (kernel_size - 1) // 2 * dilation``
+        groups (int, optional): Number of blocked connections from input channels to output channels. Default: 1
+        norm_layer (Callable[..., torch.nn.Module], optional): Norm layer that will be stacked on top of the convolution layer. If ``None`` this layer wont be used. Default: ``torch.nn.BatchNorm2d``
+        activation_layer (Callable[..., torch.nn.Module], optinal): Activation function which will be stacked on top of the normalization layer (if not None), otherwise on top of the conv layer. If ``None`` this layer wont be used. Default: ``torch.nn.ReLU``
+        dilation (int): Spacing between kernel elements. Default: 1
+        inplace (bool): Parameter for the activation layer, which can optionally do the operation in-place. Default ``True``
+        bias (bool, optional): Whether to use bias in the convolution layer. By default, biases are included if ``norm_layer is None``.
+
```
Comment (datumbox) [inline torchvision/ops/misc.py:131]: If we deprecate, we probably can remove the documentation similar to what we did below:
https://github.com/pytorch/vision/blob/0663377dda32036d3f479ee3e54e7ffcc01e513c/torchvision/models/mobilenetv3.py#L23-L24

File: torchvision/ops/misc.py [hunk 4]
```diff
@@ -114,56 +100,125 @@ def __init__(
...
+            kernel_size,
+            stride,
+            padding,
+            groups,
+            norm_layer,
+            activation_layer,
+            dilation,
+            inplace,
+            bias,
+        )
```
Comment (datumbox) [inline torchvision/ops/misc.py:167]: To avoid redefining every single parameter we could do something like:
https://github.com/pytorch/vision/blob/8a16e12f3a7f10d124b26aeb7975cd2bf4a81695/torchvision/models/quantization/inception.py#L28-L29

Then throw the warning afterwards.

Also it seems that this class shouldn't be inheriting from `_ConvNormActivation` but straights from `Conv2dNormActivation` which does exactly the same thing.

File: torchvision/ops/misc.py [hunk 5]
```diff
@@ -196,27 +251,57 @@ def __init__(
...
+
+
+class SqueezeExcitation(torch.nn.Module):
```
Comment (datumbox) [inline torchvision/ops/misc.py:271]: I assume this code is completely unmodified and Github fails to create a proper diff. Could you confirm?

**Review by datumbox — COMMENTED** [review]
@oke-aditya I've added a few suggestions that aim to simplify the code duplication. Let me know what you think. 

All of the proposals assume we will deprecate the old class. There is an alterantive but let's discuss it here after we clean up this proposal.

File: torchvision/ops/misc.py (hunk 5 above)
Comment (oke-aditya) [reply torchvision/ops/misc.py:177]: Yes, I haven't touched SqueezeExcitation

File: torchvision/ops/misc.py [hunk 6]
```diff
@@ -114,16 +100,80 @@ def __init__(
...
         ]
+
         if norm_layer is not None:
             layers.append(norm_layer(out_channels))
+
         if activation_layer is not None:
             params = {} if inplace is None else {"inplace": inplace}
             layers.append(activation_layer(**params))
         super().__init__(*layers)
-        _log_api_usage_once(self)
```
Comment (oke-aditya) [inline torchvision/ops/misc.py:123]: Where do I log? In the parent class or in each child class individually?

File: torchvision/ops/misc.py [hunk 7]
```diff
@@ -114,16 +100,80 @@ def __init__(
...
+
         self.out_channels = out_channels
 
 
+class Conv2dNormActivation(_ConvNormActivation):
+    """
+    Configurable block used for Convolution-Normalzation-Activation blocks.
+
+    Args:
+        in_channels (int): Number of channels in the input image
```
Comment (oke-aditya) [inline torchvision/ops/misc.py:120]: We need to document these both times, as we are not exposing the base class.

File: torchvision/ops/misc.py [hunk 8]
```diff
@@ -114,16 +100,80 @@ def __init__(
...
+        groups (int, optional): Number of blocked connections from input channels to output channels. Default: 1
+        norm_layer (Callable[..., torch.nn.Module], optional): Norm layer that will be stacked on top of the convolution layer. If ``None`` this layer wont be used. Default: ``torch.nn.BatchNorm2d``
+        activation_layer (Callable[..., torch.nn.Module], optinal): Activation function which will be stacked on top of the normalization layer (if not None), otherwise on top of the conv layer. If ``None`` this layer wont be used. Default: ``torch.nn.ReLU``
+        dilation (int): Spacing between kernel elements. Default: 1
+        inplace (bool): Parameter for the activation layer, which can optionally do the operation in-place. Default ``True``
+        bias (bool, optional): Whether to use bias in the convolution layer. By default, biases are included if ``norm_layer is None``.
+
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
```
Comment (datumbox) [inline torchvision/ops/misc.py:134]: Sorry, perhaps I should have been clearer. My earlier comment for using `*args: Any, **kwargs: Any` referred ONLY to the deprecated class. To the normal classes should have proper declarations as you had them. This will allow for better static analysis and IDE integration.

Note that this is also necessary for having different default values on `norm_layers` between 2d and 3d. In this implementation the default should be `torch.nn.BatchNorm2d`, not None as above.

File: torchvision/ops/misc.py (hunk 4 above)
Comment (datumbox) [reply torchvision/ops/misc.py]: This comment was referring ONLY to the deprecated method... The rest should be kept as normal. Sorry I wasn't clear.

File: torchvision/ops/misc.py [hunk 9]
```diff
@@ -114,16 +100,80 @@ def __init__(
...
+        layer = torch.nn.Conv2d
+        super().__init__(layer=layer, *args, **kwargs)
+
+
+class Conv3dNormActivation(_ConvNormActivation):
+    """
+    Configurable block used for Convolution-Normalzation-Activation blocks.
```
Comment (datumbox) [inline torchvision/ops/misc.py:141]:
```suggestion
    Configurable block used for Convolution3d-Normalzation-Activation blocks.
```

File: torchvision/ops/misc.py [hunk 10]
```diff
@@ -124,6 +112,122 @@ def __init__(
...
+        norm_layer: Optional[Callable[..., torch.nn.Module]] = torch.nn.BatchNorm2d,
+        activation_layer: Optional[Callable[..., torch.nn.Module]] = torch.nn.ReLU,
+        dilation: int = 1,
+        inplace: Optional[bool] = True,
+        bias: Optional[bool] = None,
+    ) -> None:
+
+        conv_layer = torch.nn.Conv2d
+        super().__init__(
+            in_channels,
```
Comment (oke-aditya) [inline torchvision/ops/misc.py:151]: I guess I can do something like 
`super().__init__(conv_layer=conv_layer, *args, **kwargs)`

Comment (datumbox) [reply torchvision/ops/misc.py:151]: No really. I think you are good as you are.

File: torchvision/ops/misc.py [hunk 11]
```diff
@@ -154,9 +181,36 @@ class Conv3dNormActivation(_ConvNormActivation):
...
+        padding: Optional[int] = None,
+        groups: int = 1,
+        norm_layer: Optional[Callable[..., torch.nn.Module]] = torch.nn.BatchNorm3d,
+        activation_layer: Optional[Callable[..., torch.nn.Module]] = torch.nn.ReLU,
+        dilation: int = 1,
+        inplace: Optional[bool] = True,
+        bias: Optional[bool] = None,
+    ) -> None:
+
+        conv_layer = torch.nn.Conv3d
```
Comment (datumbox) [inline torchvision/ops/misc.py:199]: Same here

**Review by datumbox — COMMENTED** [review]
Last couple of nits:

**Comment by jdsgomes** [comment]
I agree with the comments from @datumbox. I suggested otherwise initially , but the fact is that there are already 39 occurrences of  `ConvNormActivation` internally and 81 occurences of `ops.misc import ConvNormActivation` in Github (and 374 of `ConvNormActivation` which might contain similarly named blocks from other implementations). This on its own wont stop from us deprecating it, but together with the fact that is was introduced recently it does’t make sense to deprecated it. 

So @oke-aditya agrees, I would suggest to make the changes proposed by @datumbox and we can get it merged!

**Comment by jdsgomes** [comment]
@oke-aditya raised a good point offline related to this changes. This was recently exposed in [ops/__init__.py](https://github.com/pytorch/vision/blob/release/0.12/torchvision/ops/__init__.py) and we will be reverting this on a separate PR in the spirit not promoting the usage of `ConvNormActivation` directly. 

So I think I can work on this separately and create a PR that will
- remove ConvNormActivation `ops/__init__.py`
- change all imports of `ops.convNormActivation` to `ops.misc.ConvNormActivation`
- remove all mentions of `ConvNormActivation` in the docs

**Comment by oke-aditya** [comment]
So in a brief offline discussion with @jdsgomes 

We are exposing `ConvNormActivatoin` in current release branch
https://github.com/pytorch/vision/blob/release/0.12/torchvision/ops/__init__.py
@jdsgomes  will patch this :smile: 

Also, fully agreeing that PyTorch is highly stable and we need to keep our codebase backward compatible and avoid such sudden changes. (Having the power to deprecate doesn't mean we should keep using it often :smile: ) 

I will refactor the implementation to be compatible with the old one as @datumbox suggested.

File: torchvision/ops/misc.py [hunk 12]
```diff
@@ -67,21 +67,8 @@ def __repr__(self) -> str:
...
-        stride (int, optional): Stride of the convolution. Default: 1
-        padding (int, tuple or str, optional): Padding added to all four sides of the input. Default: None, in wich case it will calculated as ``padding = (kernel_size - 1) // 2 * dilation``
-        groups (int, optional): Number of blocked connections from input channels to output channels. Default: 1
-        norm_layer (Callable[..., torch.nn.Module], optional): Norm layer that will be stacked on top of the convolutiuon layer. If ``None`` this layer wont be used. Default: ``torch.nn.BatchNorm2d``
-        activation_layer (Callable[..., torch.nn.Module], optinal): Activation function which will be stacked on top of the normalization layer (if not None), otherwise on top of the conv layer. If ``None`` this layer wont be used. Default: ``torch.nn.ReLU``
-        dilation (int): Spacing between kernel elements. Default: 1
-        inplace (bool): Parameter for the activation layer, which can optionally do the operation in-place. Default ``True``
-        bias (bool, optional): Whether to use bias in the convolution layer. By default, biases are included if ``norm_layer is None``.
-
+    DEPRECATED
```
Comment (datumbox) [inline torchvision/ops/misc.py:70]: It's not deprecated. It's just a schrodinger class (public and private at the same time). 🐈‍⬛ 

Just removed the documentation and we are good.

File: torchvision/ops/misc.py [hunk 13]
```diff
@@ -124,6 +116,105 @@ def __init__(
         self.out_channels = out_channels
```
Comment (oke-aditya) [inline torchvision/ops/misc.py:117]: If I add a warning here. 
```
        warnings.warn(
            "Do not use ConvNormActivation class directly. "
            "Instead use torchvision.ops.Conv2dNormActivation instead.",
            FutureWarning,
        )
```


This will also get raised when user uses `Conv2dNormActivation` or `Conv3dNormActivation` as we are calling super() of this class.
I don't know about workaround to avoid raising such inherited warning.

File: torchvision/ops/misc.py [hunk 14]
```diff
@@ -97,13 +84,16 @@ def __init__(
         dilation: int = 1,
         inplace: Optional[bool] = True,
         bias: Optional[bool] = None,
+        conv_layer: Callable[..., torch.nn.Module] = torch.nn.Conv2d,
     ) -> None:
+
```
Comment (datumbox) [inline torchvision/ops/misc.py:89]: I believe we can detect it if someone instantiated `ConvNormActivation` directly and throw a warning.

The way to do this is by something like:
```python
if self.__class__ == ConvNormActivation:
    warn("Please dont call me directly ....")
```

File: torchvision/ops/misc.py (hunk 13 above)
Comment (datumbox) [reply torchvision/ops/misc.py:117]: I was reviewing at the same time again... Github should add a feature to let you know someone is writing comments.

See my suggestion here: https://github.com/pytorch/vision/pull/5445#discussion_r814141497

File: torchvision/ops/misc.py (hunk 12 above)
Comment (oke-aditya) [reply torchvision/ops/misc.py]: Yeah schrodinger cats are everywhere! :) Fixed

File: torchvision/ops/misc.py (hunk 14 above)
Comment (oke-aditya) [reply torchvision/ops/misc.py:89]: Nice trick! I didn't know this. Works like charm. Fixed it!

**Review by jdsgomes — COMMENTED** [review]
LGTM, thank you for working on this @oke-aditya !

File: torchvision/ops/misc.py [hunk 15]
```diff
@@ -114,15 +101,119 @@ def __init__(
...
         if activation_layer is not None:
             params = {} if inplace is None else {"inplace": inplace}
             layers.append(activation_layer(**params))
         super().__init__(*layers)
         _log_api_usage_once(self)
         self.out_channels = out_channels
 
+        if self.__class__ == ConvNormActivation:
+            warnings.warn("Don't use ConvNormActivation directly. Use Conv2dNormActivation instead.")
```
Comment (datumbox) [inline torchvision/ops/misc.py:116]:
```suggestion
            warnings.warn("Don't use ConvNormActivation directly, please use Conv2dNormActivation and Conv3dNormActivation instead.")
```

**Review by datumbox — COMMENTED** [review]
@oke-aditya  Looks good! 

The work on the block itself is complete but now we need to replace all the usages of `ConvNormActivation` from the code-base with `Conv2dNormActivation`. This update needs to happen on this PR to avoid generating unwanted warnings to our users.

**Comment by oke-aditya** [comment]
Yes I'm on it!!

**Comment by oke-aditya** [comment]
Please compare https://github.com/pytorch/vision/pull/5445/commits/9a5ab98c2566486cb608e1c41cde02c654d199eb 
and https://github.com/pytorch/vision/pull/5440/files 

I think I have replaced all the places!

**Review by datumbox — APPROVED** [review]
LGTM, thanks @oke-aditya!

@jdsgomes What you thing of cherrypicking this on the release so that we don't advertise using ConvNormActication on the mobilenets? Did the final RC was produced by the release engineering team?

**Comment by jdsgomes** [comment]
After discussing offline with @datumbox we decided not to cherrypick this one as we missed the official cut off date and this is quite a bit change.

**Review by jdsgomes — APPROVED** [review]
LGTM!
